presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java

```diff
@@ -294,6 +345,278 @@ public Page getOutput()
+                blockEncodingBuffers[i].resetBuffers();
+            }
+        }
+
+        private void setupDecodedBlocksAndPositions(DecodedObjectNode[] decodedObjectNodes)
+        {
+            int channelCount = decodedObjectNodes.length;
+            for (int i = 0; i < channelCount; i++) {
+                blockEncodingBuffers[i].setupDecodedBlocksAndPositions(decodedObjectNodes[i], positions, positionCount);
+            }
+        }
+
+        private void calculateRowSizes(boolean containsNonFixedWidthBlocks, int rowSizeForfixedWidthBlocks, ArrayList<Integer> variableWidthBlockChannels)
+        {
+            if (!containsNonFixedWidthBlocks) {
+                return;
+            }
+
+            ensureRowSizesCapacity();
+
+            resetRowSizes();
+
+            for (int variableWidthChannel : variableWidthBlockChannels) {
+                blockEncodingBuffers[variableWidthChannel].accumulateRowSizes(rowSizes);
+            }
+
+            for (int i = 0; i < positionCount; i++) {
+                rowSizes[i] += rowSizeForfixedWidthBlocks;
+            }
+        }
+
+        private void resetRowSizes()
+        {
+            Arrays.fill(rowSizes, 0, positionCount, 0);
+        }
+
+        private int calculateNumberRowsFitInBuffers(boolean containsNonFixedWidthBlocks, int rowSizeForfixedWidthBlocks, int startPosition)
+        {
+            // TODO: validate rows and rowSizes length >= numberOfRows > firstRowIndex
+            // Note that the capacity limit may be exceeded by 1 row.
+            int remainingBytes = capacity - getBuffersSizeInBytes();
+
+            if (!containsNonFixedWidthBlocks) {
+                //int rowsFit = (int) ceil(((double) remainingBytes / rowSizeForfixedWidthBlocks));
+                int rowsFit = max(remainingBytes / rowSizeForfixedWidthBlocks, 1);
+                bufferFull = (rowsFit <= positionCount - startPosition);
+                return min(rowsFit, positionCount - startPosition);
+            }
+
+            verify(rowSizes != null);
+            for (int i = startPosition; i < positionCount; i++) {
+                int newRemainingBytes = remainingBytes - rowSizes[i];
+
+                if (newRemainingBytes <= 0) {
+                    bufferFull = true;
+                    return max(i - startPosition, 1);
+                }
+                remainingBytes = newRemainingBytes;
+            }
+
+            return positionCount - startPosition;
+        }
+
+        private void flush(OutputBuffer outputBuffer)
+        {
+            if (isBufferEmpty()) {
+                return;
+            }
+
+            Slice slice = serialize();
+            SerializedPage serializedPage = serde.wrapBuffer(slice, bufferedRowCount);
+            outputBuffer.enqueue(lifespan, partition, Collections.singletonList(serializedPage));
+            pagesAdded.incrementAndGet();
+            rowsAdded.addAndGet(bufferedRowCount);
```

Comment:
Original PagePartitioner used rowsAdded and appendRow. Since we separated out the two operators, I think it's better to keep them as is. If we change the names in OptimizedPartitionedOutputOperator, we may also want to change the original PartitionedOutputOperator, which seems unnecessary.